File: open-telemetry/course/04-api-sdk-architecture/01-approaches/java/src/main/java/com/course/ApiNoOp.java
package com.course;

import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.Tracer;

public class ApiNoOp {

    public static void main(String[] args) {
        System.out.println("=== Running Java with API ONLY (No SDK Registered) ===");

        // GlobalOpenTelemetry returns a DefaultTracer (No-Op) if no SDK is configured
        Tracer tracer = GlobalOpenTelemetry.getTracer("library-sample", "1.0.0");

        Span span = tracer.spanBuilder("library_operation").startSpan();
        try {
            System.out.println("Span isRecording: " + span.isRecording());
            System.out.println("Is valid span context: " + span.getSpanContext().isValid());
            System.out.println("Trace ID: " + span.getSpanContext().getTraceId());

            span.setAttribute("data.processed", 100L);
            span.addEvent("completed");
        } finally {
            span.end();
        }

        System.out.println("=== Result: No output, zero overhead, zero errors! ===\n");
    }
}

File: open-telemetry/course/04-api-sdk-architecture/01-approaches/python/api_noop.py
from opentelemetry import trace

def main():
    print("=== Running with API ONLY (No SDK Registered) ===")
    
    # 1. Obtain a tracer directly from the global API
    # Since no TracerProvider was registered, OTel automatically returns a DefaultTracer (No-Op)
    tracer = trace.get_tracer("library-sample", "1.0.0")

    # 2. Start span
    with tracer.start_as_current_span("library_operation") as span:
        print(f"Span is_recording: {span.is_recording()}")
        
        # Attributes are safely accepted but dropped immediately in-memory
        span.set_attribute("data.processed", 100)
        span.add_event("completed")

        ctx = span.get_span_context()
        print(f"Is valid trace context: {ctx.is_valid}")
        print(f"Trace ID format: {trace.format_trace_id(ctx.trace_id)}")

    print("=== Result: Zero overhead, zero output, zero crashes! ===\n")

if __name__ == "__main__":
    main()
